### PR TITLE
Turns on the history scavenger for SQL backends

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
In this PR, I enabled the history scanner to delete unused history tree branches. In addition, I removed an else which made it impossible for the history scanner and task queue scanner to be on simultaneously (task queue scanner would take priority).

**Why?**
I made these changes for https://github.com/temporalio/temporal/issues/3419 in order to rule this out as an option for why disk usage is increasing so much over time for SQL backends.

**How did you test it?**
I added a new test to the Scanner which verifies that we actually start the scavenger now regardless of the storage backend we're using.

**Potential risks**
The biggest risk I can think of is that this change could turn on the scavenger for NoSQL backends, and it could have a bug which causes it to remove branches which it shouldn't, which could interfere with workflow progress and lead to data loss.

**Is hotfix candidate?**
This PR is not a hot fix candidate which requires a notification to the broader community. However, we should tell the people on the original PR that this fix is in, so that they can patch it and see if it fixes their issue.

### CODE REVIEW

File: service/worker/service.go [hunk 1]
```diff
@@ -449,17 +456,21 @@ func (s *Service) startBatcher() {
 	}
 }
 
-func (s *Service) startScanner() {
-	sc := scanner.New(
+func (s *Service) initScanner() {
+	s.scanner = scanner.New(
```
Comment (yiminc) [inline service/worker/service.go:460]: What is this change for?

File: service/worker/scanner/scanner.go [hunk 2]
```diff
@@ -137,20 +143,25 @@ func (s *Scanner) Start() error {
...
+
+	if s.context.cfg.HistoryScannerEnabled() {
+		s.wg.Add(1)
 		go s.startWorkflowWithRetry(historyScannerWFStartOptions, historyScannerWFTypeName)
 		workerTaskQueueNames = append(workerTaskQueueNames, historyScannerTaskQueueName)
 	}
 
 	for _, tl := range workerTaskQueueNames {
-		work := worker.New(s.context.sdkSystemClient, tl, workerOpts)
+		work := s.context.workerFactory.New(s.context.sdkSystemClient, tl, workerOpts)
```
Comment (yiminc) [inline service/worker/scanner/scanner.go:164]: what is the difference here?

File: service/worker/service.go (hunk 1 above)
Comment (MichaelSnowden) [reply service/worker/service.go:460]: This allows us to close the scanner when the service closes (which waits for all of its coroutines to finish)

File: service/worker/scanner/scanner.go (hunk 2 above)
Comment (MichaelSnowden) [reply service/worker/scanner/scanner.go:164]: Injecting a worker factory here allows me to test which workers were created. If I just use `worker.New`, it does some assertion that the worker is created by a real sdk client.

File: service/worker/scanner/scanner.go [hunk 3]
```diff
@@ -84,13 +86,15 @@ type (
...
 		historyClient    historyservice.HistoryServiceClient
+		workerFactory    sdk.WorkerFactory
 	}
 
 	// Scanner is the background sub-system that does full scans
 	// of database tables to cleanup resources, monitor anamolies
 	// and emit stats for analytics
 	Scanner struct {
 		context scannerContext
+		wg      sync.WaitGroup
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/worker/scanner/scanner.go:97]: Some of my assertions are for side-effects in goroutines. If I don't wait for them to finish, the test could be flaky due to race conditions.

File: service/worker/scanner/scanner.go [hunk 4]
```diff
@@ -40,13 +41,14 @@ import (
...
 	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/sdk"
 
 	"go.temporal.io/server/common/backoff"
 	"go.temporal.io/server/common/dynamicconfig"
 	"go.temporal.io/server/common/log/tag"
 )
 
-const (
+var (
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/worker/scanner/scanner.go:51]: This allows me to override the sleep duration to 0 during testing.

[the remark this answers is not in the document]
Comment (yux0) [reply service/worker/scanner/scanner.go:51]: maybe we can remove this delay and leverage the retry

File: service/worker/scanner/scanner_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+func TestScanner(t *testing.T) {
+	suite.Run(t, new(scannerTestSuite))
+}
+
+func (s *scannerTestSuite) TestScannerEnabled() {
+	defer func(originalDelay time.Duration) {
+		scannerStartUpDelay = originalDelay
+	}(scannerStartUpDelay)
```
Comment (yux0) [inline service/worker/scanner/scanner_test.go:54]: will this work after you set the delay to 0?

Comment (MichaelSnowden) [reply service/worker/scanner/scanner_test.go:54]: Yes. All of the deferred function's arguments are computed immediately. For example:

```
package main

func main() {
	x := 0
	defer func(x int) {
		println(x) // prints 0
	}(x)
	x = 42
}
```

File: service/worker/scanner/scanner.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/worker/scanner/scanner.go]: Going to rerun the CI with that idea

**Review by yux0 — APPROVED** [review]

File: service/worker/scanner/scanner.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/worker/scanner/scanner.go]: The CI does pass with this removed. I think it's probably safe to do

**Pull request merged**